src/query_ctrl.ts

```diff
@@ -72,27 +80,45 @@ export class CogniteQueryCtrl extends QueryCtrl {
     this.isAllSelected =
       this.target.assetQuery.timeseries &&
       this.target.assetQuery.timeseries.every(ts => ts.selected);
+
+    appEvents.on(failedResponseEvent, this.handleError);
+    appEvents.on(datapointsLimitWarningEvent, this.handleWarning);
   }
 
+  handleWarning = ({ refId, warning }: QueryDatapointsLimitWarning) => {
+    if (this.target.refId !== refId) {
+      return;
+    }
+
+    this.target.warning = warning;
+  };
+
+  handleError = ({ refId, error }: QueryRequestError) => {
+    if (refId !== this.target.refId) {
+      return;
+    }
+
+    this.target.error = error;
+  };
+
   getOptions(query: string, type: string) {
     return this.datasource.getOptionsForDropdown(query || '', type).then(options => {
       _.defer(() => this.$scope.$digest()); // need to force the update on the dropdown
       return options;
     });
   }
 
-  onChangeCustom() { // todo: why it doesn't work without this hack?
-    try {
-      parse(this.target.expr);
-      this.target.error = '';
-      this.onChangeInternal();
-    } catch (e) {
-      this.target.error = e.message;
-    }
+  apply() {
+    this.refresh(); // Asks the panel to refresh data.
```

Comment:
guess we had additional logic before, it can be replaced with direct call.
But we lose readability at the same time, cause inside AngularJS template we're calling function which is not in the controller (it inherited from extended class, but grafana doesn't document base `QueryCtrl` at all)

Mb we should leave it in favor of readability?

Reply:
yep, ok. but maybe rename to refreshFields or something?

Reply:
yep